src/dstack/_internal/server/services/gateways/__init__.py

```diff
@@ -254,26 +260,70 @@ async def register_service_jobs(
         raise ServerClientError("Gateway has no instance associated with it")
 
     domain = gateway.wildcard_domain.lstrip("*.") if gateway.wildcard_domain else None
-    private_bytes, public_bytes = generate_rsa_key_pair_bytes(comment=f"{project.name}/{run_name}")
 
     job.job_spec.gateway.gateway_name = gateway.name
-    job.job_spec.gateway.ssh_key = private_bytes.decode()
     if domain is not None:
         job.job_spec.gateway.secure = True
         job.job_spec.gateway.public_port = 443
         job.job_spec.gateway.hostname = f"{run_name}.{domain}"
     else:
-        job.job_spec.gateway.secure = False
-        # use provided public port
-        job.job_spec.gateway.hostname = gateway.gateway_compute.ip_address
+        raise ServerClientError("Domain is required for gateway")
 
     await run_async(
-        configure_gateway_over_ssh,
-        f"ubuntu@{gateway.gateway_compute.ip_address}",
+        _gateway_preflight,
+        project.name,
+        job.job_spec.gateway.hostname,
         gateway.gateway_compute.ssh_private_key,
-        public_bytes.decode(),
-        [job],
+        project.ssh_private_key,
+        job.job_spec.gateway.options,
+    )
+
+
+def _gateway_preflight(
+    project: str,
+    domain: str,
+    gateway_ssh_private_key: str,
+    project_ssh_private_key: str,
+    options: dict,
+):
+    try:
+        with gateway_tunnel_client(domain, gateway_ssh_private_key) as client:
+            client.preflight(project, domain, project_ssh_private_key, options)
+    except requests.RequestException:
+        raise GatewayError(f"Gateway is not working")
+
+
+async def update_gateways(session: AsyncSession):
+    if os.environ.get("DSTACK_SKIP_GATEWAY_UPDATE", None):
```

Comment:
Put DSTACK_SKIP_GATEWAY_UPDATE in server settings so that envs are not scattered around.